### ISSUE TITLE
Possible memory leak

### ISSUE THREAD

**Issue by IOTech17** [cloudradar-monitoring/rport#801]
Hi,

I am hosting rport server in a docking container using this image : 
https://github.com/IOTech17/rport-docker

I can see that the memory usage since version 0.9.0 is going up without ever coming down : 

![Screenshot from 2022-11-21 22-11-11](https://user-images.githubusercontent.com/68947790/203163430-2dd00472-dd5a-4a10-9292-543c1e53330c.png)

It seems that the process associated with the connection are not killed when the connection is closed.

I had the memory reach more than 6GB before it started having issue.

When restarting the container it start back at 40MB

Do you have any knowledge of this issue?

**Comment by thorstenkramm** [comment]
@IOTech17 Thanks for reporting. This kind of feedback is essential for us.
How many clients are connected to your rport server?

Finding the root cause of issues is done often by exclusion procedure or trail and error.
Do you get the same memory consumption behaviour if you run rport outside the container?

This is the memory graph of our test server with ~800 clients connected.
<img width="1631" alt="image" src="https://user-images.githubusercontent.com/16852023/203262576-8cf5dccf-ebc7-4a68-ba21-906d319a0634.png">

And this is the memory consumption of the rportd process.
<img width="1644" alt="image" src="https://user-images.githubusercontent.com/16852023/203263043-078cc49c-d25a-425c-a3f3-3d8118d6f86f.png">

What is your graph showing? Virt or Res memory. Having an increase in virtual memory is normal. This memory is freed when needed. Real consumption is “resident memory”.

**Comment by IOTech17** [comment]
@thorstenkramm 

My env works this way :

I am using Proxmox Hypervisor to host FedoraCore OS VM and then deploy the docker container on FedoraCore OS.

As you can see on the graph the memory consumption is slowly increasing overtime even without anybody using the Rport.

![Screenshot from 2022-11-22 10-24-11](https://user-images.githubusercontent.com/68947790/203277681-dbcc44ab-24aa-42e1-a729-418287491600.png)


On this instance I have 21 computer connected (Windows/Linux/Mac OS).

I never face this issue prior to the version 0.9.0

I can only the realtime memory usage but it doesn't tell me if it is cache or the current memory usage.

I think it is the current memory usage as it start having issue when reach a certain memory usage.

I will add htop inside the docker to track what exactly is happening.

**Comment by thorstenkramm** [comment]
@IOTech17  Definitely no! Rportd is a single process daemon. This is not normal and maybe related to Docker. 
We have never seen this before.

**Comment by thorstenkramm** [comment]
@IOTech17 A tunnel does not create a new process. And looking at the pid column, we can see that you have multiple rportd processes running. Maybe the way the container starts the process is not optimal. Remove any settings regarding auto start of a process, restart the container, and start the rportd process manually. Do multiple processes in the container come up? If not, you know where the problem is. 

As a side note: Why do you run such a complex setup? Promox has native LXD container support. In my opinion, running a VM with Fedora to run rport inside a container there is not needed. When using LXD containers, you can use systemd to keep rport up and running.

**Comment by IOTech17** [comment]
@thorstenkramm I cannot run the process manually as I am using supervisord to run 2 services : guacd to access rdp via web browser and the rportd server.

I will try to disable supervisord ans see what is happening.

I am using multiple fcos VM to host mutliples containers and managing them using portainer.

**Comment by IOTech17** [comment]
same thing happen if I just run rportd command from the docker-compose.

![Uploading Screenshot from 2022-11-22 12-21-58.png…]()

**Comment by thorstenkramm** [comment]
> Is there a way to use an other container to run guacd? In the documentation you recommend running guacd on localhost, but could it be used with another VM/container?

Yes. Guacd can run anywhere. You can use the [official Guacd docker container](https://hub.docker.com/r/guacamole/guacd).

**Comment by IOTech17** [comment]
But would rport support it?

Last time I tried to use an external container for guacd I could not use RDP via Browser

**Comment by thorstenkramm** [comment]
@IOTech17 
> But would rport support it?

Yes. You can enter any address into `rportd.conf` for the guacd. It's not limited to localhost.

**Comment by IOTech17** [comment]
@thorstenkramm
Thanks!

I will try to see if it helps.

Let me investigate the issue on my side.

I would really recommend working on supporting rport server inside a docker container, it makes the deployment much more easier

**Comment by IOTech17** [comment]
@thorstenkramm from the official rport technical documentation : 

Apache Guacamole Proxy Server 1.4.X has to be running on 127.0.0.1
RPort server configuration tunnel_proxy_cert_file and tunnel_proxy_key_file must be set up. The guacd proxy depends on the generic built-in TLS reverse proxy, which is enabled by providing a key and a certificate.

**Comment by thorstenkramm** [comment]
Disk read/write on an 800-clients-rportd

![image](https://user-images.githubusercontent.com/16852023/203810865-79b97371-9bee-4153-b477-79eef915688c.png)

It depends mostly on your monitoring settings. The more monitoring data you collect, the more disk writes.

**Comment by thorstenkramm** [comment]
Disk read/write on an 800-clients-rportd

<img width="1631" alt="image" src="https://user-images.githubusercontent.com/16852023/203811477-a41b57c3-768a-4490-a394-3aacb7d48e86.png">

It depends mostly on your monitoring settings. The more monitoring data you collect, the more disk writes.

**Comment by IOTech17** [comment]
@thorstenkramm 

So, it seems that I was right, there is a memory leak.

I am using top and can see only one process running for rport.

But the ram slowly climb without ever going down.

![Screenshot from 2022-11-24 21-30-09](https://user-images.githubusercontent.com/68947790/203859730-14137eb7-4987-4ffd-bb98-3d94d025d6dd.png)

**Comment by thorstenkramm** [comment]
@thorstenkramm Thanks for all your testing.
Can you share a graph over a longer period of time?
This is how it looks like on one of our systems (currently 800 clients connected)
<img width="1631" alt="image" src="https://user-images.githubusercontent.com/16852023/203950817-dd1ca240-e10a-4b42-be74-ac5f172e5516.png">
The increase in memory usage is caused by the increasing number of connected clients.

**Comment by IOTech17** [comment]
![Screenshot from 2022-12-07 15-48-45](https://user-images.githubusercontent.com/68947790/206742572-95d3904d-7f7f-44bd-8932-e74afc7332b3.png)

**Comment by thorstenkramm** [comment]
@thorstenkramm 
600MB is not healthy. 
Below, you can see the memory consumption on a vanilla Ubuntu 20.04 (without docker) and 1000 clients connected to rport.
<img width="1397" alt="image" src="https://user-images.githubusercontent.com/16852023/207357246-057e14f9-0b50-4d26-8a35-7668d135a335.png">

**Comment by IOTech17** [comment]
@thorstenkramm thank you for confirming.

what is the recommended OS to run the rport server?

Something is wrong with the memory consumption but I do not know where it is coming from, but when I check TOP it is rportd using the memory.

**Comment by thorstenkramm** [comment]
@IOTech17 We recommend Debian 11 or Ubuntu 22.04 because you have packages for Guacamole. 
If you don't need RDP via browser, any RedHat-based distro is fine too.
